tools/mk_vendor

```diff
@@ -0,0 +1,854 @@
+        map find_build                      |
+        map normalize_path                  |
+        awk '!seen[$0]++'
+}
+
+read_build_flags() {
+    local config=${1-}
+
+    toml_get -a "${config}" vendor.build   |
+        map replace "build.@@.cmake_flags" |
+        map toml_get -a "${config}"        |
+        remove_duplicated_flags
+
+    toml_get -a "${config}" vendor.build    |
+        map replace "build.@@.cflags_extra" |
+        map toml_get -a -d, "${config}"     |
+        map replace "-DCFLAGS_EXTRA=@@"
+
+    toml_get -a "${config}" vendor.build     |
+        map replace "build.@@.ldflags_extra" |
+        map toml_get -a -d, "${config}"      |
+        map replace "-DLDFLAGS_EXTRA=@@"
+}
+
+read_build_options() {
+    local config=${1-}
+
+    get_all() {
+        toml_get -a "${config}" vendor.build   |
+            map replace "build.@@.options"     |
+            map toml_get -a "${config}"        |
+            awk '!seen[$0]++'
+    }
+
+    IFS=$'\n' read -r -d '' -a build_options < <(toml_get -a "${config_file}" vendor.options && printf '\0')
+
+    while IFS=$'\n' read -r option_name
+    do
+        if [[ -z ${option_name} ]]; then
+            continue
+        fi
+
+        if is_in build_options "${option_name}"; then
+            printf '%s=ON\n' "${option_name}"
+        else
+            printf '%s=OFF\n' "${option_name}"
+        fi
+    done < <(get_all) | to_upper | map replace '-DWITH_@@'
+}
+
+toml_get() {
+    local config=""
+    local search=""
+    local separator=$'\n'
+    local is_array=FALSE
+    local is_non_empty=FALSE
+
+    while [[ $# -gt 2 ]]
+    do
+        case $1 in
+          -a)
+            is_array=TRUE
+            ;;
+          -n)
+            is_non_empty=TRUE
+            ;;
+          -d*)
+            separator=${1#-d}
+            ;;
+          *)
+            ;;
+        esac
+        shift
+    done
+
+    config="${1-}"
+    search="${2-}"
+
+    # check if this search key was overriden
+    local override
+    override=$(to_upper "${search}")_OVERRIDE
+
+    local value
+    if [[ ${!override-NOT_SET} != NOT_SET ]]; then
+        value=${!override}
+    else
+        value="$(toml get -r "${config}" "${search}" || true)"
+
+        # remove surrounding brackets if present
+        if [[ ${is_array} = TRUE ]]; then
+            value=${value#'['}
+            value=${value%']'}
+        fi
+    fi
+
+    if [[ ${is_non_empty} = TRUE ]] && [[ -z ${value} ]]; then
+        die "configuration key '${search}' is not set or empty"
+    fi
+
+    if [[ ${is_array} = FALSE ]] || [[ -z "${value}" ]]; then
+        printf '%s\n' "${value}"
+        return 0
+    fi
+
+    IFS=',' read -r -a array <<< "${value}"
+
+    # remove quotes from each array element
+    IFS=$'\n' read -r -d '' -a array < <((printf '%s\n' "${array[@]}" | unquote) && printf '\0')
+
+    value=$(IFS="${separator}"; printf '%s\n' "${array[*]}")
+
+    printf "%s\n" "${value}"
+    return 0
+}
```

Comment:
This seems very manual and easier to do in most other languages.

Reply:
Welcome to bash 😄